Add translation support to menu

kanishk98 opened the pull request:
<!--
Remove the fields that are not appropriate
Please include:
-->

**What's this PR do?**

Gets the `default_language` of the user's realm and updates the app menu accordingly. 
Also updates the `google-translate-api` module used to generate the `.json` files in `app/translations/` to prevent `BAD_REQUEST` errors.

**Any background context you want to provide?**

Discussion at #407. 
Please also see #437 and #436. 

**Screenshots?**

![link](https://user-images.githubusercontent.com/24617297/56419994-e6abc780-62b9-11e9-903e-65daad3b991d.gif)


**You have tested this PR on:**
  - [x] Windows
  - [ ] Linux/Ubuntu
  - [ ] macOS

**Further work**

I'll also be pushing changes to the settings page once we decide that this approach is fine.

kanishk98 committed:
i18n: Add new values to translations

akashnimare commented:
Let's have a discussion about this in https://chat.zulip.org/#narrow/stream/16-desktop/topic/spellchecker.

abhigyank commented:
@kanishk98 can you rebase this?

kanishk98 commented:
@abhigyank sorry, missed this last night. Rebased the PR.

kanishk98 commented:
After some users reporting bugs with their JSON files recently, I decided to take another look at this PR to try and reduce the dependence of this feature on the disk. The last commit broke functionality, so marking this as WIP for now. Will just move back to the disk-based approach if nothing works out. :)

kanishk98 commented:
Alright, so the JSON db errors mostly stemmed from switching tabs and now that we've addressed that, the PR works as expected for me. I think this is ready for review again. Removing WIP flag.

akashnimare commented:
Can you fix the merge conflicts?

kanishk98 committed:
i18n: Setup module and add translation-util.js

* Add i18n package.
* Use system locale for TranslationUtil.

kanishk98 committed:
locales: Change translate API to handle rate limit

Also add catch block to log possible errors.

kanishk98 committed:
i18n: Add translations to app menu

* Uses new keys added to locale-template.json
* Disregards org names

akashnimare commented on app/renderer/js/main.js:
```diff
@@ -621,6 +630,20 @@ class ServerManagerView {
+			const index = this.getTabIndex(domain);
+			const oldLanguage = this.tabs[index].props.language;
+			this.tabs[index].props.language = serverLanguage;
+			if (oldLanguage !== serverLanguage) {
```
This is to handle the case where both org have the same language, right?

kanishk98 replied:
Uh, yeah, pretty much. More accurately, this is for the case when the app default (English) is not the same as the server language.

akashnimare commented:
LGTM. @priyank-p quick review?

priyank-p commented on app/renderer/js/utils/translation-util.js:
```diff
@@ -0,0 +1,31 @@
+	constructor() {
+		if (instance) {
+			return this;
+		} else {
```
Minor nit: you can just remove this else clause here.

priyank-p commented on app/renderer/js/utils/translation-util.js:
```diff
@@ -0,0 +1,31 @@
+				register: this
+			});
+		}
+		return this;
```
Remove this! This is basically useless.

priyank-p commented on app/main/menu.js:
```diff
@@ -441,44 +464,58 @@ class AppMenu {
 				role: 'selectall'
 			}]
 		}, {
-			label: '&View',
```
Why did we have `&` here for?

kanishk98 replied:
Not sure. Maybe @akashnimare can help out here.

akashnimare replied:
This came from a standard electron template. Not sure if we need this now.

priyank-p reviewed (commented):
LGTM! Left some minor comments

kanishk98 commented:
Pushed the suggested changes.

kanishk98 committed:
i18n: Change menu according to system locale.

kanishk98 committed:
i18n: Add translations for general settings.

kanishk98 committed:
i18n: Add translations for network settings.

kanishk98 committed:
i18n: Add translations for connected orgs.

kanishk98 committed:
i18n: Add translations for shortcuts pane.

kanishk98 committed:
i18n: Add translations to settings sidebar.

kanishk98 committed:
i18n: Add translations to the new server form.

akashnimare commented on app/renderer/js/pages/preference/add-certificate.js:
```diff
@@ -18,12 +19,12 @@ class AddCertificate extends BaseComponent {
 			<div class="settings-card certificates-card">
 				<div class="certificate-input">
-					<div>Organization URL</div>
+					<div>${TranslationUtil.__('Organization URL')}</div>
```
@kanishk98 instead of `TranslationUtil` you can use `t` maybe?

priyank-p commented:
@kanishk98 I rebased this PR to solve merge conflicts related to typescript can you cross check?

kanishk98 commented:
> I rebased this PR to solve merge conflicts related to typescript can you cross check?

I did, everything looks great. Thanks a ton, @priyank-p!

> @kanishk98 instead of TranslationUtil you can use t maybe?

@akashnimare made the changes and force pushed. 
I think this PR is ready for merging now. Let me know if any other changes are needed.

akashnimare commented:
@kanishk98 can you update this whenever you get time?

kanishk98 commented:
Fixed merge conflicts and updated PR.

akashnimare commented on app/renderer/js/utils/translation-util.ts:
```diff
@@ -0,0 +1,35 @@
+import electron = require('electron');
+import i18n = require('i18n');
+
+let instance: TranslationUtil = null;
```
@priyank-p, @vsvipul is this correct? We have used following at other places -
```
let instance: null | ConfigUtil = null;
```

priyank-p replied:
Yes, it is! We need to do a sweep in couple of places.

akashnimare commented on app/renderer/js/utils/translation-util.ts:
```diff
@@ -0,0 +1,35 @@
+	}
+
+	__(phrase: string): string {
+		return i18n.__({ phrase, locale: app.getLocale() });
```
@kanishk98 should we also set `ENGLISH` here as a fallback?

kanishk98 replied:
I doubt that it's required because `app.getLocale()` should always work, but it's never bad to be safe, so I added a commit for that anyway. We can drop that if others feel it is unnecessary.

kanishk98 replied:
I'm assuming you're talking about the corner case where `app.getLocale()` returns a falsey value.

akashnimare replied:
Correct.

akashnimare commented:
@vsvipul can you test this on Linux? This works fine for me on macOS.

akashnimare commented:
@kanishk98 merge the conflicts. I think it's GTG and we could include this in the next stable release.

kanishk98 commented:
@akashnimare fixed the conflicts. I also tested it on Linux, so the PR looks good.

kanishk98 committed:
i18n: Use English as fallback language.

In case app.getLocale() returns a falsey value, English is used as the
app language.

akashnimare commented:
@kanishk98 thanks for working on this.
